### ISSUE TITLE
/web/image controller returns images with session_id Set-Cookie header, making it hard to use a public cache for images and other semi-static content

### ISSUE THREAD

**Issue by LeartS** [odoo/odoo#17084]
Go to runbot and access /shop, see the requests from the developer console of your browser.

You'll see a bunch of GET for the (resized) product images, like:

    GET http://229693-10-0-b75326.runbot11.odoo.com/web/image/product.template/436/image/300x300?unique=d90082a

Inspecting the reponse headers, you'll see:

```
Server: nginx/1.10.0
Date: Thu, 18 May 2017 17:33:18 GMT
Connection: keep-alive
X-Content-Type-Options: nosniff
Etag: "e5a6e7107da61bc93007b3a6cf8ee609"
Cache-Control: max-age=604800
Set-Cookie: session_id=ea2912212330d879c6aab1d0e9ce4999ee7a4955; Expires=Wed, 16-Aug-2017 17:33:18 GMT; Max-Age=7776000; Path=/
```

For some reason, every image has `Set-Cookie` in the headers setting the `session_id`. I don't see why a simple GET of a public image should set the session in the cookie; plus this makes it hard to use a public cache for e.g. product images in the ecommerce; because responses with Set-Cookie cannot be cached transparently.

Is there a reason for the Set-Cookie, and/or is there documentation/examples of using a public cache for public images that bypasses the cookie header problem?

**Comment by crazy-projects** [comment]
In my case is even worse, not only the Set-cookie header is sent, but also the Cache-control max-age is set to 0 (efectively preventing any caching):
```
cache-control: max-age=0
content-length: 6771
content-type: image/jpeg
date: Sun, 17 Dec 2017 06:11:09 GMT
etag: "9104f4431e4c994b1916594f37e38a10"
server: Werkzeug/0.9.6 Python/2.7.9
set-cookie: session_id=<session id> GMT; Max-Age=7776000; Path=/
x-content-type-options: nosniff
```
@LeartS For the time being, you can work arround this by creatin a location for the images (/web/image/*) and using the directives "proxy_hide_header set-cookie".

For the cache control, you have to use proxy_hide_header, proxy_ignore_header, so nginx doesn't forwards nor obeys the faulty cahce control header, AND user proxy_cache_valid so nginx knows how long should it cache the objects

**Comment by pedrobaeza** [comment]
@JKE-be I think this is interesting for you.

**Comment by lem8r** [comment]
@crazy-projects cache-control tag is zero because web editor is not adding `unique` tag to image url.
This can easily be fixed changing two lines in odoo's code https://github.com/odoo/odoo/pull/26039/files
And it is not so easy to fix with custom module because `ImageDialog` is not exported.
Possible workaround is to override cache-control tag at nginx level.

**Comment by crazy-projects** [comment]
@lem8r Yes, the Nginx workarround works fine, you have to set-up arround 5 paths and override its headers (the set-cookie and cache control), and that is what i'm doing right now (even witout nginx, you can do it with cloudfront)

However, the ideal solution would be odoo setting the correct headers, since who knows better than odoo the content it's serving and it's "cacheability"? that way the nginx config is much easier.

PS: thanks for the patch, i'll implement it in my instance.

**Comment by brunocasado** [comment]
@crazy-projects did you solved the problem? ( about max age)

**Comment by JKE-be** [comment]
Thanks for path

@lem8r Can you give more info on how your patch will solve smthing ? 
You cannot use only your patch, you need anyway to remove header no?

Btw, your uniq hash will be not updated when image will be updated, so seems not correct or I miss smthing.


Btw, Imo it is normal to use nginx before Odoo, and if you install website, use cdn for all /web/image t is not a luxury.

**Comment by lem8r** [comment]
@JKE-be 

> @lem8r Can you give more info on how your patch will solve smthing ?

Sure. Adding unique to image url will change how image controller will behave.
`www.odoo.com/web/image/902051` will deliver image with max-age: 0
`www.odoo.com/web/image/902051-bc17e6a` will deliver image with max-age: 604800


> You cannot use only your patch, you need anyway to remove header no?

Currently [header is correctly set](https://github.com/odoo/odoo/blob/b81c2bce846c94b5695d680ebbc30f1cfe51e614/odoo/addons/base/models/ir_http.py#L374) if unique is provided.


> Btw, your uniq hash will be not updated when image will be updated, so seems not correct or I miss smthing.

Yeah, good point. I take checksum of attachment as a value for unique. web_editor will store final link to template. In case when image will be updated in some other place (not web_editor) template will not be recomputed. But how it is possible to change content of the image in attachment while keeping it's id?
As far as I understand  we assume that we can edit html contend with web_editor only. (except removing attachment directly). We can do better job with web_editor by storing image's url as a field value which is computed each time template is rendered. In that case we have some overhead but we will know real checksum every time.

> Btw, Imo it is normal to use nginx before Odoo, and if you install website, use cdn for all /web/image t is not a luxury.

For CDN we still have to provide correct headers.

**Comment by psy0rz** [comment]
So this is still not fixed in odoo13 it seems?

Odoo is setting cookies on all static resources it seems?

max-age=604800 seems to be set as well so thats nice at least.
